src/script/src/python/engine.rs

```diff
@@ -20,27 +20,157 @@ use std::task::{Context, Poll};
+use common_query::prelude::Signature;
 use common_query::Output;
 use common_recordbatch::error::{ExternalSnafu, Result as RecordBatchResult};
 use common_recordbatch::{RecordBatch, RecordBatchStream, SendableRecordBatchStream};
-use datatypes::schema::SchemaRef;
+use datafusion_expr::Volatility;
+use datatypes::schema::{ColumnSchema, SchemaRef};
+use datatypes::vectors::VectorRef;
 use futures::Stream;
 use query::QueryEngineRef;
 use session::context::QueryContext;
 use snafu::{ensure, ResultExt};
 use sql::statements::statement::Statement;
 
 use crate::engine::{CompileContext, EvalContext, Script, ScriptEngine};
-use crate::python::coprocessor::{exec_parsed, parse, CoprocessorRef};
+use crate::python::coprocessor::{exec_parsed, parse, AnnotationInfo, CoprocessorRef};
 use crate::python::error::{self, Result};
 
 const PY_ENGINE: &str = "python";
 
+#[derive(Debug)]
+pub struct PyUDF {
+    copr: CoprocessorRef,
+}
+
+impl std::fmt::Display for PyUDF {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}({})->",
+            &self.copr.name,
+            &self.copr.deco_args.arg_names.join(",")
+        )
+    }
+}
+
+impl PyUDF {
+    fn from_copr(copr: CoprocessorRef) -> Arc<Self> {
+        Arc::new(Self { copr })
+    }
+
+    /// Register to `FUNCTION_REGISTRY`
+    fn register_as_udf(zelf: Arc<Self>) {
+        FUNCTION_REGISTRY.register(zelf)
+    }
+    fn register_to_query_engine(zelf: Arc<Self>, engine: QueryEngineRef) {
+        engine.register_function(zelf)
+    }
+
+    /// Fake a schema, should only be used with dynamically eval a Python Udf
+    fn fake_schema(&self, columns: &[VectorRef]) -> SchemaRef {
+        let arg_names = &self.copr.deco_args.arg_names;
+        let col_sch: Vec<_> = columns
+            .iter()
+            .enumerate()
+            .map(|(i, col)| ColumnSchema::new(arg_names[i].to_owned(), col.data_type(), true))
+            .collect();
+        let schema = datatypes::schema::Schema::new(col_sch);
+        Arc::new(schema)
+    }
+}
+
+impl Function for PyUDF {
+    fn name(&self) -> &str {
+        &self.copr.name
+    }
+
+    fn return_type(
+        &self,
+        _input_types: &[datatypes::prelude::ConcreteDataType],
+    ) -> common_query::error::Result<datatypes::prelude::ConcreteDataType> {
+        // TODO(discord9): use correct return annotation if exist
+        match self.copr.return_types.get(0) {
+            Some(Some(AnnotationInfo {
+                datatype: Some(ty), ..
+            })) => Ok(ty.to_owned()),
+            _ => PyUdfSnafu {
+                msg: "Can't found return type for python UDF {self}",
+            }
+            .fail(),
+        }
+    }
+
+    fn signature(&self) -> common_query::prelude::Signature {
+        // try our best to get a type signature
+        let mut arg_types = Vec::with_capacity(self.copr.arg_types.len());
+        let mut know_all_types = true;
```

Comment:
I think it can be replaced by `Iterator#all(F)` function.

https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.all